common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java

```diff
@@ -0,0 +1,220 @@
+                                Logger.info(TAG, "An instance of PivProvider was removed from Security static list upon YubiKey device connection being closed.");
+                            } else {
+                                Telemetry.emit(pivProviderStatusEvent.putPivProviderRemoved(false));
+                                Logger.info(TAG, "An instance of PivProvider was not present in Security static list upon YubiKey device connection being closed.");
+                            }
+                            startDiscoveryCallback.onClosedConnection();
+                        }
+
+                    });
+                }
+            }
+        });
+    }
+
+    /**
+     * Ceases usb discovery for YubiKeys.
+     * Usually called when a host fragment is destroyed.
+     */
+    @Override
+    public void stopDiscovery() {
+        mYubiKitManager.stopUsbDiscovery();
+    }
+
+    /**
+     * Attempt to get a PivSession instance in order to carry out methods
+     *  implemented in YubiKitSmartcardSession.
+     * @param callback Contains callbacks to run when a PivSession is successfully instantiated and when any exception is thrown due to a connection issue.
+     */
+    @Override
+    public void attemptDeviceSession(@NonNull final ISessionCallback callback) {
+        final String methodTag = TAG + "attemptDeviceSession:";
+        synchronized (sDeviceLock) {
+            if (mDevice == null) {
+                Logger.error(methodTag, MDEVICE_NULL_ERROR_MESSAGE, null);
+                callback.onException(new Exception());
+            }
+            //Request a connection from mDevice so that we can get a PivSession instance.
+            mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+                @Override
+                public void invoke(@NonNull final Result<UsbSmartCardConnection, IOException> value) {
+                    try {
+                        final SmartCardConnection c = value.getValue();
+                        final PivSession piv = new PivSession(c);
+                        YubiKitSmartcardSession session = new YubiKitSmartcardSession(piv);
+                        callback.onGetSession(session);
+                    } catch (final Exception e) {
+                        callback.onException(e);
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Checks if a YubiKey is currently connected.
+     * @return true if YubiKey is currently connected. Otherwise, false.
+     */
+    @Override
+    public boolean isDeviceConnected() {
+        synchronized (sDeviceLock) {
+            return mDevice != null;
+        }
+    }
+
+    /**
+     * Adds a PivProvider instance to the Java static Security List (and emits relevant telemetry).
+     */
+    @Override
+    public void prepareForAuth() {
```

Comment:
Suggested above -- this method naming seems a bit assumptive about the use of the smartcard. What if we're just signing something using its cert? Is that auth?

Suggest naming this more generically, along the lines of `initialize()` or similar

Reply:
I created `PrepareForAuth` because some logic needed to happen (specifically, adding an implemented security provider to the static Java Security list) **right before** `proceed` is called on the `ClientCertRequest`.  I'm not certain that `initialize()` would be specific enough here since I wouldn't want someone to think it should be called any time other than before `proceed`. Would `initializeBeforeProceedingWithRequest()` be an ok rename? It's kinda long, but at least more accurate than 'prepareForAuth' :)

Reply:
Ended up renaming to `initBeforeProceedingWithRequest`. Chatted with Brian offline regarding the purpose of this method, and after thinking about it for a bit, I think it could still be useful for future implementations, as long as the developer makes sure to clean up well in `onDestroy` or the `onClosedConnection` callback. 
I also think it's more organized to separate the PivProvider logic from the `getKeyForAuth` method in order to avoid hiding the PivProvider logic in a semi related method. 
I'll probably revisit this in the future, especially when we add support for more types of smartcards.